pkg/schedule/schedulers/balance_range.go

```diff
@@ -167,6 +230,286 @@ func newBalanceRangeScheduler(opController *operator.Controller, conf *balanceRa
+		}
+
+		if err != nil {
+			balanceRangeCreateOpFailCounter.Inc()
+			return nil
+		}
+		sourceLabel := strconv.FormatUint(sourceID, 10)
+		targetLabel := strconv.FormatUint(targetID, 10)
+		op.FinishedCounters = append(op.FinishedCounters,
+			balanceDirectionCounter.WithLabelValues(s.GetName(), sourceLabel, targetLabel),
+		)
+		op.SetAdditionalInfo("sourceScore", strconv.FormatInt(plan.sourceScore, 10))
+		op.SetAdditionalInfo("targetScore", strconv.FormatInt(plan.targetScore, 10))
+		op.SetAdditionalInfo("tolerate", strconv.FormatInt(plan.tolerate, 10))
+		return op
+	}
+	balanceRangeNoReplacementCounter.Inc()
+	return nil
+}
+
+// balanceRangeSchedulerPlan is used to record the plan of balance key range scheduler.
+type balanceRangeSchedulerPlan struct {
+	sche.SchedulerCluster
+	// stores is sorted by score desc
+	stores []*core.StoreInfo
+	// scoreMap records the storeID -> score
+	scoreMap     map[uint64]int64
+	source       *core.StoreInfo
+	sourceScore  int64
+	target       *core.StoreInfo
+	targetScore  int64
+	region       *core.RegionInfo
+	fit          *placement.RegionFit
+	averageScore int64
+	job          *balanceRangeSchedulerJob
+	opInfluence  operator.OpInfluence
+	tolerate     int64
+}
+
+func (s *balanceRangeScheduler) prepare(cluster sche.SchedulerCluster, opInfluence operator.OpInfluence, job *balanceRangeSchedulerJob) (*balanceRangeSchedulerPlan, error) {
+	filters := s.filters
+	switch job.Engine {
+	case core.EngineTiKV:
+		filters = append(filters, filter.NewEngineFilter(string(types.BalanceRangeScheduler), filter.NotSpecialEngines))
+	case core.EngineTiFlash:
+		filters = append(filters, filter.NewEngineFilter(string(types.BalanceRangeScheduler), filter.SpecialEngines))
+	default:
+		return nil, errs.ErrGetSourceStore.FastGenByArgs(job.Engine)
+	}
+	sources := filter.SelectSourceStores(cluster.GetStores(), filters, cluster.GetSchedulerConfig(), nil, nil)
+	if sources == nil {
+		return nil, errs.ErrStoresNotEnough.FastGenByArgs("no store to select")
+	}
+
+	krs := core.NewKeyRanges(job.Ranges)
+	scanRegions, err := cluster.BatchScanRegions(krs)
+	if err != nil {
+		return nil, err
+	}
+
+	// storeID <--> score mapping
+	scoreMap := make(map[uint64]int64, len(sources))
+	for _, source := range sources {
+		scoreMap[source.GetID()] = 0
+	}
+	totalScore := int64(0)
+	for _, region := range scanRegions {
+		for _, peer := range job.Role.getPeers(region) {
+			scoreMap[peer.GetStoreId()] += 1
+			totalScore += 1
+		}
+	}
+
+	sort.Slice(sources, func(i, j int) bool {
+		role := job.Role
+		iop := role.getStoreInfluence(opInfluence.GetStoreInfluence(sources[i].GetID()))
+		jop := role.getStoreInfluence(opInfluence.GetStoreInfluence(sources[j].GetID()))
+		iScore := scoreMap[sources[i].GetID()]
+		jScore := scoreMap[sources[j].GetID()]
+		return iScore+iop > jScore+jop
+	})
+
+	averageScore := int64(0)
+	if len(sources) != 0 {
```

Comment:
This has been checked in L390. Maybe we can remove this.

Reply:
done